core/src/banking_stage/transaction_scheduler/multi_iterator_consume_scheduler.rs

```diff
@@ -0,0 +1,617 @@
+        // completed.
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 2);
+        let (thread_0_work, thread_0_ids) = collect_work(&work_receivers[0]);
+        assert_eq!(thread_0_ids, [txids!([0])]);
+        assert_eq!(collect_work(&work_receivers[1]).1, [txids!([1])]);
+
+        // Cannot schedule even on next pass because of lock conflicts
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 0);
+
+        // Complete batch on thread 0. Remaining tx can be scheduled onto thread 1
+        scheduler.complete_batch(thread_0_work[0].batch_id, &thread_0_work[0].transactions);
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 1);
+
+        assert_eq!(collect_work(&work_receivers[1]).1, [txids!([2])]);
+    }
+
+    #[test]
+    fn test_schedule_priority_guard() {
+        let (mut scheduler, work_receivers) = create_test_frame(2);
+
+        let accounts = (0..6).map(|_| Keypair::new()).collect_vec();
+        let mut container = create_container([
+            (&accounts[0], vec![accounts[1].pubkey()], 1, 3),
+            (&accounts[2], vec![accounts[3].pubkey()], 1, 2),
+            (
+                &accounts[1],
+                vec![accounts[2].pubkey(), accounts[4].pubkey()],
+                1,
+                1,
+            ),
+            (&accounts[4], vec![accounts[5].pubkey()], 1, 0),
+        ]);
+
+        // high priority transactions [0, 1] do not conflict, and should be
+        // scheduled to *different* threads.
+        // low priority transaction [2] conflicts with both, and thus will
+        // not be schedulable until one of the previous transactions is
+        // completed.
+        // low priority transaction [3] does not conflict with any scheduled
+        // transactions, but the priority guard should stop it from taking
+        // a lock that transaction [2] needs.
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 2);
+        let (thread_0_work, thread_0_ids) = collect_work(&work_receivers[0]);
+        assert_eq!(thread_0_ids, [txids!([0])]);
+        assert_eq!(collect_work(&work_receivers[1]).1, [txids!([1])]);
+
+        // Cannot schedule even on next pass because of lock conflicts
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 0);
+
+        // Complete batch on thread 0. Remaining txs can be scheduled onto thread 1
+        scheduler.complete_batch(thread_0_work[0].batch_id, &thread_0_work[0].transactions);
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 2);
+
+        assert_eq!(
+            collect_work(&work_receivers[1]).1,
+            [txids!([2]), txids!([3])]
+        );
```

Comment:
`2` does conflict with both `0` and `1`.
The reason `2` cannot be scheduled, is because `0` is on thread 0, and `1` is on thread 1.
If either completes, then `2` can be queued behind either transaction `0` or `1` on the same thread - since queued work is executed serially.

In other words, once `0` is completed (accounts unlocked). `2` only conflicts with a single thread, therefore it is schedulable on that thread. `3` only conflicts with `2`, so it is also schedulable on that thread.

Reply:
Ah right! thanks for explaining.